cpp/src/search/contains_nested.cu

```diff
@@ -59,4 +65,173 @@ bool contains_nested_element(column_view const& haystack,
   });
 }
 
+namespace {
+/**
+ * @brief Convert an index `i` iterating in reverse order in the range `[-1, -size-1)` into the
+ *        valid index range `[0, size)`.
+ *
+ * This conversion is needed when we want to use two strong index types but are limited to use
+ * only one type. As such, we can use non-negative indices to denote one strong index type and
+ * negative indices to denote the other. After recognizing the corresponding type, we need to
+ * convert the negative indices into their original values.
+ */
+struct index_adapter {
```

Comment:
Does the `index_adapter` need to be a functor or can it just be a device function? I don't see it being used in any cases where a functor is required.

Reply:
Good point. I'll make it `constexpr auto func()`.